test/e2e/auth/node_authn.go

```diff
@@ -0,0 +1,130 @@
+	"time"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/master/ports"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthenticator]", func() {
+
+	f := framework.NewDefaultFramework("node-authn")
+	var ns string
+	var nodeIPs []string
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(nodeList.Items)).NotTo(BeZero())
+
+		pickedNode := nodeList.Items[0]
+		// Here we only need to care about Internal IP, since the pods running in the cluster
+		// can see the internal addresses.
+		nodeIPs = framework.GetNodeAddresses(&pickedNode, v1.NodeInternalIP)
+
+		// make sure ServiceAccount admission controller is enabled, so secret generation on SA creation works
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Get("default", metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(sa.Secrets)).NotTo(BeZero())
+	})
+
+	It("The kubelet's main port 10250 should reject requests with no credentials", func() {
+		pod := createNodeAuthTestPod(f)
+		for _, nodeIP := range nodeIPs {
+			// Anonymous authentication is disabled by default
+			result, err := framework.LookForStringInPodExec(pod.Namespace,
+				pod.Name,
+				[]string{"curl",
+					"-s",
+					"-I",
+					"--insecure",
+					fmt.Sprintf("https://%s:%v/metrics", nodeIP, ports.KubeletPort),
+				},
+				"", // get the result for later comparison
+				time.Minute)
+			Expect(err).NotTo(HaveOccurred())
+
+			if !strings.Contains(result, "401") && !strings.Contains(result, "403") {
+				Expect(fmt.Errorf("the kubelet's main port 10250 should reject requests with no credentials, but got: %s", result)).To(HaveOccurred())
+			}
+		}
+	})
+
+	It("The kubelet can delegate ServiceAccount tokens to the API server", func() {
+		By("create a new ServiceAccount for authentication")
+		trueValue := true
+		newSA := &v1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      "node-auth-newSA",
+			},
+			AutomountServiceAccountToken: &trueValue,
+		}
+		_, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Create(newSA)
+		Expect(err).NotTo(HaveOccurred())
+
+		pod := createNodeAuthTestPod(f)
+
+		for _, nodeIP := range nodeIPs {
+			result, err := framework.LookForStringInPodExec(pod.Namespace,
+				pod.Name,
+				[]string{"sh",
+					"-c",
+					fmt.Sprintf("curl -sI --insecure --header \"Authorization: Bearer `%s`\" %s",
+						"cat /var/run/secrets/kubernetes.io/serviceaccount/token",
+						fmt.Sprintf("https://%s:%v/metrics", nodeIP, ports.KubeletPort)),
+				},
+				"", // get the result for later comparison
+				time.Minute)
+			Expect(err).NotTo(HaveOccurred())
+
+			if !strings.Contains(result, "403") && !strings.Contains(result, "401") {
```

Comment:
`Expect(result).To(And(BeNumerically(">=", 200), BeNumerically("<", 300)), ...)`

Reply:
I don't particularly care for the gomega DSL, so feel free to do these checks explicitly, but semantically this is the condition you should check for.

Reply:
You might be able to simplify this to just: `Expect(result).To(Equal("200"), ...)`

Reply:
Here we're using a new SA token when sending requests to the kubelet. The expected result should be 403, and authentication is okay.

> Expect(result).To(Equal("200"), ...)

@tallclair I don't quite understand where `200` comes.